Content.Server/Power/EntitySystems/ChargerSystem.cs

```diff
@@ -13,69 +14,69 @@ internal sealed class ChargerSystem : EntitySystem
+        SubscribeLocalEvent<ChargerComponent, ComponentStartup>(OnStartup);
         SubscribeLocalEvent<ChargerComponent, PowerChangedEvent>(OnPowerChanged);
         SubscribeLocalEvent<ChargerComponent, EntInsertedIntoContainerMessage>(OnInserted);
         SubscribeLocalEvent<ChargerComponent, EntRemovedFromContainerMessage>(OnRemoved);
         SubscribeLocalEvent<ChargerComponent, ContainerIsInsertingAttemptEvent>(OnInsertAttempt);
         SubscribeLocalEvent<ChargerComponent, ExaminedEvent>(OnChargerExamine);
     }
 
+    private void OnStartup(EntityUid uid, ChargerComponent component, ComponentStartup args)
+    {
+        UpdateStatus(uid, component);
+    }
+
     private void OnChargerExamine(EntityUid uid, ChargerComponent component, ExaminedEvent args)
     {
         args.PushMarkup(Loc.GetString("charger-examine", ("color", "yellow"), ("chargeRate", component.ChargeRate)));
     }
 
     public override void Update(float frameTime)
     {
-        foreach (var comp in EntityManager.EntityQuery<ChargerComponent>())
+        foreach (var comp in EntityManager.EntityQuery<ChargerActiveComponent>())
         {
-            comp.OnUpdate(frameTime);
-        }
-    }
-    private void OnChargerInit(EntityUid uid, ChargerComponent component, ComponentInit args)
-    {
-        _itemSlotsSystem.AddItemSlot(uid, "charger-slot", component.ChargerSlot);
-    }
+            if (!TryComp(comp.Owner, out ChargerComponent? charger))
+                return;
 
-    private void OnChargerRemove(EntityUid uid, ChargerComponent component, ComponentRemove args)
-    {
-        _itemSlotsSystem.RemoveItemSlot(uid, component.ChargerSlot);
-    }
+            if (!_itemSlotsSystem.TryGetSlot(comp.Owner, charger.SlotId, out ItemSlot? slot))
+                continue;
+
+            if (charger.Status == CellChargerStatus.Empty || charger.Status == CellChargerStatus.Charged || !slot.HasItem)
+                continue;
 
+            TransferPower(comp.Owner, charger, frameTime);
+        }
+    }
+    
     private void OnPowerChanged(EntityUid uid, ChargerComponent component, ref PowerChangedEvent args)
     {
-        component.UpdateStatus();
+        UpdateStatus(uid, component);
     }
 
     private void OnInserted(EntityUid uid, ChargerComponent component, EntInsertedIntoContainerMessage args)
     {
         if (!component.Initialized)
             return;
 
-        if (args.Container.ID != component.ChargerSlot.ID)
+        if (args.Container.ID != component.SlotId)
             return;
 
-        // try get a battery directly on the inserted entity
-        if (!TryComp(args.Entity, out component.HeldBattery))
-        {
-            // or by checking for a power cell slot on the inserted entity
-            _cellSystem.TryGetBatteryFromSlot(args.Entity, out component.HeldBattery);
-        }
-
-        component.UpdateStatus();
+        AddComp<ChargerActiveComponent>(uid);
+        UpdateStatus(uid, component);
```

Comment:
AddComp should be done in UpdateStatus as it might be adding the component even if there's nothing to do from UpdateStatus earlying out.